### PR TITLE
Global object visibility fix

### PR DESCRIPTION
**Pull request by jimil749**
Solve for #1472 .

GDevelop

Sorry for the quality, but I hope it does the needful :)

### CODE REVIEW

**Comment by jimil749** [comment]
@4ian can you please review?

File: newIDE/app/src/MainFrame/index.js [hunk 1]
```diff
@@ -1602,6 +1603,14 @@ class MainFrame extends React.Component<Props, State> {
     editorTab.editorRef.updateToolbar();
   }
 
+  forceUpdateObjectsList() {
+    const editorTab = getCurrentTab(this.state.editorTabs);
+    if (
+      editorTab.key.startsWith('layout ' + editorTab.name)
```
Comment (4ian) [inline newIDE/app/src/MainFrame/index.js]: This is risky as a change in this key later could silently break this. Can you think of another way of doing this?

Comment (Bouh) [reply newIDE/app/src/MainFrame/index.js]: I've was an idea in a previous commit, but not used.
Add a props type for earch type of tab editor.
This is an idea, maybe another way existing, for get the type without this addition.
https://github.com/4ian/GDevelop/pull/1429/commits/27331594e878eed7aae543128286888e86d53c2a

Comment (jimil749) [reply newIDE/app/src/MainFrame/index.js]: Well yes, it had no provision to know the type of the editor, hence had to use this work-around as I think key uniquely determines each type of editor! :) 

Thanks @Bouh for the idea, I'll try and implement it.

**Comment by Bouh** [comment]
Another way can be reload the editor when the state change from active false to true.
Like this only the editor used is refresh :)

**Comment by 4ian** [comment]
> Another way can be reload the editor when the state change from active false to true.
Like this only the editor used is refresh :)

Would be great to try this. I'm not a big fan of adding a "type" prop, because it's also fragile. For example, you wrote "eventsEditor" instead of "EventsEditor". This leads to hard to debug code.

Prefer if possible a solution where you don't give the responsibility of knowing what to refresh to MainFrame. MainFrame is just here to "host" the editors. We don't want it to be responsible to look at them and asking them to refresh if they are a particular type, or if blabla or if blabla... the poor MainFrame is already too big and underpaid. It's doing too much.

Instead, give the responsibility to the editor: MainFrame can tell them "please refresh" when they are activated (i.e: shown on screen). Then it's the editor responsibility to implement a refresh method that will update what is needed - some won't do anything, others will do stuff.
Better: this can be adapted to any other editor in the future without changing a single line of code in MainFrame!

**Comment by jimil749** [comment]
I did not get everything of that. 
So, do you mean to say that we need a method in MainFrame which checks if an editor is acivated and upon being active it will ask/call a method in respective editor to perform the refresh?
This way we will have add a refresh method in each editor. Some won't do anything as of now (return null) and some (SceneEditor here) will update the required component. This way the refresh is performed by the respective editors and not the MainFrame
 Is this how it should be?

**Comment by jimil749** [comment]
Also we know when a tab editor is being shown to the screen (_onEditorTabActive) does that. So maybe we can use that to perform a refresh for the editors?

**Comment by 4ian** [comment]
> So maybe we can use that to perform a refresh for the editors?

Yes that sounds like a good idea!

**Comment by jimil749** [comment]
Added a performRefresh() function for each editor, which is called when an editor is being activated. It can perform updates as per the editor requirements. Here, it just performs the refresh of objectList inside the SceneEditor

**Comment by jimil749** [comment]
@4ian can you check?

**Comment by 4ian** [comment]
I check once a day, prefer to ping me if no answer after more than 24 hours :)

File: newIDE/app/src/MainFrame/Editors/DebuggerEditor.js [hunk 2]
```diff
@@ -17,6 +17,10 @@ export default class DebuggerEditor extends BaseEditor {
     if (this.editor) this.editor.updateToolbar();
   }
 
+  performRefresh() {
+    return null;
```
Comment (4ian) [inline newIDE/app/src/MainFrame/Editors/DebuggerEditor.js]: Let's remove the `return null` and just write `// No updates to be done. `

File: newIDE/app/src/MainFrame/index.js [hunk 3]
```diff
@@ -1611,6 +1611,7 @@ class MainFrame extends React.Component<Props, State> {
 
   _onEditorTabActive = (editorTab: EditorTab) => {
     this.updateToolbar();
+    editorTab.editorRef.performRefresh();
```
Comment (4ian) [inline newIDE/app/src/MainFrame/index.js]: nitpicking: can we call this `forceUpdateEditor`?

"forceUpdate" is a hint at the forceUpdate from React. Refresh has a slightly different meaning (refresh would be something like making network requests to "refresh" your profile).

Comment (4ian) [reply newIDE/app/src/MainFrame/index.js]: Let's also add here a comment why we're doing this:
`// Ensure the editors shown on screen are updated. This is for example useful if global objects have been updated in another editor.`

**Comment by 4ian** [comment]
This approach looks good! Added a few nitpickings :)

File: newIDE/app/src/MainFrame/index.js (hunk 3 above)
Comment (jimil749) [reply newIDE/app/src/MainFrame/index.js]: Will do that straight away! :)

**Comment by 4ian** [comment]
Tests are failing because Flow is signaling a potential bug: 
image

I'm fixing this but be sure to look at it next time :)

**Comment by jimil749** [comment]
I'm sorry, was a bit too careless. Will take care next time on. Thanks for the changes though! :)

**Comment by 4ian** [comment]
Thanks for your work on this :)

**Pull request merged**